### PR TITLE
created user entity, added dto, service, repository, controller amd m…

### PR DESCRIPTION
**Pull request by nastasiiaKulinich**
…apper for it

### CODE REVIEW

File: src/main/java/com/example/bookstore/controller/AuthenticationController.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import org.springframework.validation.annotation.Validated;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/auth")
+@Validated
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/controller/AuthenticationController.java]:
```suggestion
```

File: src/main/java/com/example/bookstore/model/User.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import jakarta.persistence.Table;
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id = ?")
+@Where(clause = "is_deleted=false")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/model/User.java]:
```suggestion
@SQLRestriction(value = "is_deleted=false")
```
`@Where` is deprecated

File: src/main/java/com/example/bookstore/service/user/UserServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+@RequiredArgsConstructor
+@Service
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/service/user/UserServiceImpl.java]:
```suggestion
        if (userRepository.existsByEmail(requestDto.getEmail())) {
```
It's better approach

File: src/main/java/com/example/bookstore/service/user/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+            throw new RegistrationException("Can`t register user");
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/service/user/UserServiceImpl.java]:
```suggestion
            throw new RegistrationException(String.format("User with this email: %s already exists", request.email()));
```

File: src/main/java/com/example/bookstore/service/user/UserServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        }
+        User user = new User();
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/service/user/UserServiceImpl.java]:
```suggestion
        User user = userMapper.toUser(request);
```

File: src/main/java/com/example/bookstore/validation/FieldMatchValidator.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+    @Override
+    public void initialize(FieldMatch constraintAnnotation) {
+        firstField = constraintAnnotation.firstField();
+        secondField = constraintAnnotation.secondField();
+    }
+
+    @Override
+    public boolean isValid(Object value, ConstraintValidatorContext context) {
+        try {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/validation/FieldMatchValidator.java]:
```suggestion
        Object firstValue = new BeanWrapperImpl(value).getPropertyValue(this. firstField);
        Object secondValue = new BeanWrapperImpl(value).getPropertyValue(this. secondField);
        return Objects.equals(firstValue, secondValue);
```

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+              - column:
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: email
+                  type: varchar(255)
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml:18]: Sizes should be the same as in DTO, check all

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: src/main/java/com/example/bookstore/service/user/UserServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        if (userRepository.existsByEmail(requestDto.getEmail())) {
+            throw new RegistrationException(String.format("User with this email: %s already exists",
+                    requestDto.getEmail()));
+        }
+        User user = userMapper.toModel(requestDto);
+        user.setEmail(requestDto.getEmail());
+        user.setPassword(requestDto.getPassword());
+        user.setFirstName(requestDto.getFirstName());
+        user.setLastName(requestDto.getLastName());
+        user.setShippingAddress(requestDto.getShippingAddress());
```
Comment (okuzan) [inline src/main/java/com/example/bookstore/service/user/UserServiceImpl.java]: boilerplate code - it's mapper's job

File: src/main/java/com/example/bookstore/model/User.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+import jakarta.persistence.Table;
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id = ?")
+@SQLRestriction(value = "is_deleted=false")
```
Comment (okuzan) [inline src/main/java/com/example/bookstore/model/User.java]:
```suggestion
@SQLDelete(sql = "UPDATE users SET is_deleted = TRUE WHERE id = ?")
@SQLRestriction(value = "is_deleted = FALSE")
```

File: src/main/java/com/example/bookstore/controller/AuthenticationController.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/auth")
+public class AuthenticationController {
+    private final UserService userService;
+
+    @PostMapping("/registration")
+    public UserResponseDto register(@RequestBody @Valid UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
```
Comment (okuzan) [inline src/main/java/com/example/bookstore/controller/AuthenticationController.java]: make it unchecked, and remove from method definition here and in service. If user submitted incorrect data we can't do anything about it anyway

**Review by okuzan — CHANGES_REQUESTED** [review]

**Review by okuzan — APPROVED** [review]
Well done!

**Pull request merged**